v2/apiserver/internal/core/events.go

```diff
@@ -537,35 +537,59 @@ func (e *eventsService) CancelMany(
 		)
 	}
 
-	events, err := e.eventsStore.CancelMany(ctx, selector)
+	eventCh, errCh, err := e.eventsStore.CancelMany(ctx, selector)
 	if err != nil {
 		return result, errors.Wrap(err, "error canceling events in store")
 	}
 
-	result.Count = int64(len(events.Items))
-
-	// TODO: This could take a while, so we don't do it synchronously. But what if
-	// the process dies while this is in-progress? Can we find a quicker, more
-	// efficient way to do this?
-	go func() {
-		for _, event := range events.Items {
-			if err := e.substrate.DeleteWorkerAndJobs(
-				context.Background(), // Deliberately not using request context
-				project,
-				event,
-			); err != nil {
-				log.Println(
-					errors.Wrapf(
-						err,
-						"error deleting event %q worker and jobs from the substrate",
-						event.ID,
-					),
-				)
+	// When an event comes in via the corresponding channel,
+	// kick off a go routine to execute cleanup.
+	// Otherwise, handle errors and/or finish and return once all
+	// events have come through.
+	for {
+		select {
+		case event, ok := <-eventCh:
+			if ok {
+				result.Count++
+				// TODO: This could take a while, so we don't do it synchronously. But
+				// what if the process dies while this is in-progress? Can we find a
+				// quicker, more efficient way to do this?
+				go func() {
+					if err := e.substrate.DeleteWorkerAndJobs(
+						context.Background(), // deliberately not using ctx
+						project,
+						event,
+					); err != nil {
+						log.Println(errors.Wrapf(
+							err,
+							"error deleting event %q worker and jobs from the substrate",
+							event.ID,
+						))
+					}
+				}()
+			} else {
+				// eventCh was closed, but want to keep looping through this select
+				// in case there are pending errors on the errCh still. nil channels are
+				// never readable, so we'll just nil out eventCh and move on.
+				eventCh = nil
+			}
+		case err, ok := <-errCh:
+			if ok {
+				return result, err
```

Comment:
I'm not sure we should be returning here. Do we really want to stop all further attempts at cleaning up the substrate because there was one error?

Might be better to just log it.

If we're only going to log it, let's log the error in the store and get rid of the `errCh` entirely.

If we get rid of the `errCh`, this `select` becomes a lot simpler because you don't have to go to lengths to exhaust both channels before exiting the loop.

Reply:
Thanks, updated.